aten/src/ATen/native/TensorShape.cpp

```diff
@@ -94,130 +239,52 @@ std::vector<Tensor> broadcast_tensors(TensorList tensors) {
-    cat_dim_size += tensor.sizes()[dim];
-
-    if (!tensor.is_contiguous(first_tensor_mem_format)) {
-      allContiguous = false;
-    }
-
-    if (tensor.sizes() != notSkippedTensor.sizes() ||
-        tensor.strides() != notSkippedTensor.strides()) {
-      reuse_iterator = false;
-    }
-    if (tensor.dtype() != notSkippedTensor.dtype()) {
-      no_type_promotion = false;
-    }
-  }
-  // compute the size of the result
-  auto result_size = notSkippedTensor.sizes().vec();
-  result_size[dim] = cat_dim_size;
-
-  // skip resizing if size of result is same as expected
-  // raise a warning while resizing if output has one or more elements
-  // See https://github.com/pytorch/pytorch/pull/62560#discussion_r687363362
-  // for understanding why at::native::resize_output is not called directly.
-  // if (at::native::resize_output_check(result, result_size)) {
-  // TODO: restore the above, see https://github.com/pytorch/pytorch/issues/64709
-
-  if (result.sizes() != result_size) {
-    result.resize_(result_size, first_tensor_mem_format);
+TORCH_IMPL_FUNC(cat_out_cpu)
+(ITensorList tensors,
+ int64_t dim,
+ int64_t valid,
+ bool all_contiguous,
+ bool all_same_dtype,
+ bool all_same_sizes_and_stride,
+ MemoryFormat memory_format,
+ const Tensor& result) {
+  // This is true only if there are no valid tensors (i.e. we skipped all of them)
+  if (static_cast<size_t>(valid) == tensors.size()) {
+    return;
   }
 
   if (result.numel() == 0) {
-    return result;
+    return;
   }
 
   // fast path for single thread when both inputs and result are contiguous and not empty
-  allContiguous = allContiguous && result.is_contiguous(first_tensor_mem_format);
   bool use_serial_kernel = result.numel() < at::internal::GRAIN_SIZE || at::get_num_threads() == 1;
-  ScalarType dtype = notSkippedTensor.scalar_type();
+  ScalarType dtype = tensors[valid].scalar_type();
   bool serial_dtype = (dtype == ScalarType::Double || dtype == ScalarType::Float || dtype == ScalarType::BFloat16);
-  if (use_serial_kernel && allContiguous && no_type_promotion && serial_dtype) {
+  if (use_serial_kernel && all_contiguous && all_same_dtype && serial_dtype) {
     cat_serial_stub(kCPU, result, tensors, dim);
-    return result;
+    return;
   }
 
   int64_t offset = 0;
-  if (reuse_iterator &&
-      result.is_contiguous(first_tensor_mem_format) &&
-      no_type_promotion) {
-    const auto& source_slice = notSkippedTensor;
+  if (all_same_sizes_and_stride && result.is_contiguous(memory_format) &&
+      all_same_dtype) {
+    const auto& source_slice = tensors[valid];
     auto slice_dim_size = source_slice.sizes()[dim];
     auto result_slice = result.narrow(dim, 0, slice_dim_size);
     auto result_slice_data = result_slice.data_ptr();
     auto result_stride_bytes = result.stride(dim) * elementSize(result.scalar_type());
 
     auto iter = TensorIteratorConfig()
-      .set_check_mem_overlap(false)  // Already checked above
```

Comment:
"Already checked above" still applies, right? You're doing the overlap check explicitly in the meta function

Reply:
I thought leaving the comment would be a little confusing, since we do it in META (not "above", in the same function). Maybe I could just add: "Already checked in the META function". What do you think?

Reply:
That sounds good!